login_server/src/cpp/JSONInterface/JsonUpdateUserInfos.cpp

```diff
@@ -144,32 +144,75 @@ Poco::JSON::Object* JsonUpdateUserInfos::handle(Poco::Dynamic::Var params)
 				}
+				std::string value_str = value.toString();
+				if (!value_str.size()) {
+					jsonErrorsArray.add("User.password is empty");
+				}
 				else {
-					NotificationList errors; 
-					if (!sm->checkPwdValidation(value.toString(), &errors, LanguageManager::getInstance()->getFreeCatalog(LANG_EN))) {
-						jsonErrorsArray.add("User.password isn't valid");
-						jsonErrorsArray.add(errors.getErrorsArray());
+					std::string old_password;
+					auto old_password_obj = updates->get("User.password_old");
+					if (old_password_obj.isEmpty()) {
+						jsonErrorsArray.add("User.password_old not found");
+					}
+					else if (!old_password_obj.isString()) {
+						jsonErrorsArray.add("User.password_old isn't a string");
 					}
 					else {
-						auto result_new_password = user->setNewPassword(value.toString());
-						
-						switch (result_new_password) {
-							// 0 = new and current passwords are the same
-						case 0: jsonErrorsArray.add("new password is the same as old password"); break;
-							// 1 = password changed, private key re-encrypted and saved into db
-						//case 1: extractet_values++; break;
-							// 2 = password changed, only hash stored in db, couldn't load private key for re-encryption
-						case 2: jsonErrorsArray.add("password changed, couldn't load private key for re-encryption");  break;
-							// -1 = stored pubkey and private key didn't match
-						case -1: jsonErrorsArray.add("stored pubkey and private key didn't match"); break;
+						old_password_obj.convert(old_password);
+					}
+
+					bool old_password_valid = false;
+					NotificationList errors;
+					if (old_password.size()) 
+					{
+						if (!sm->checkPwdValidation(old_password, &errors, LanguageManager::getInstance()->getFreeCatalog(LANG_EN))) {
+							jsonErrorsArray.add("User.password_old didn't match");
+							Poco::Thread::sleep(ServerConfig::g_FakeLoginSleepTime);
 						}
-						
-						
+						else 
+						{
+							auto secret_key = user->createSecretKey(old_password);
+							if (secret_key->getKeyHashed() == user_model->getPasswordHashed()) {
+								old_password_valid = true;
+							}
+							else if (secret_key.isNull()) {
+								jsonErrorsArray.add("Password calculation for this user already running, please try again later");
+							}
+							else {
+								jsonErrorsArray.add("User.password_old didn't match");
+							}
+						}
+
 					}
+					if (old_password_valid) 
+					{
+						if (!sm->checkPwdValidation(value.toString(), &errors, LanguageManager::getInstance()->getFreeCatalog(LANG_EN))) {
+							jsonErrorsArray.add("User.password isn't valid");
+							jsonErrorsArray.add(errors.getErrorsArray());
+						}
+						else 
+						{
+							auto result_new_password = user->setNewPassword(value.toString());
+
+							switch (result_new_password) {
+								// 0 = new and current passwords are the same
+							case 0: jsonErrorsArray.add("new password is the same as old password"); break;
```

Comment:
This condition is not checked on pw reset. Why here?

Reply:
I don't know what you mean. 
Which condition?
What did you mean with pw reset?
This will be used on password reset and password change.

Reply:
the condition that you set the same password. My test case on reset password always involves the same password and that works without a problem. Here you force for a different password. I don't mind, but should there not be identical policies for passwords?